File: python/lib/engine/engine_client.py
import argparse
import socket
from ..game.poker_moves import FoldAction, CallAction, CheckAction, RaiseAction
from ..game.poker_state import GameState, TerminalState, RoundState
from ..game.poker_state import STARTING_STACK, BIG_BLIND, SMALL_BLIND
from ..base.base_bot import BaseBot


class EngineClient:
    def __init__(self, pokerbot, socketfile):
        self.pokerbot = pokerbot
        self.socketfile = socketfile

    def receive(self):
        while True:
            packet = self.socketfile.readline().strip().split(' ')
            if not packet:
                break
            yield packet

    def send(self, action):
        if isinstance(action, FoldAction):
            code = 'F'
        elif isinstance(action, CallAction):
            code = 'C'
        elif isinstance(action, CheckAction):
            code = 'K'
        else:  # isinstance(action, RaiseAction)
            code = 'R' + str(action.amount)
        self.socketfile.write(code + '\n')
        self.socketfile.flush()

    def run(self):
        game_state = GameState(0, 0., 1)
        round_state = None
        active = 0
        round_flag = True
        for packet in self.receive():
            for clause in packet:
                if clause[0] == 'T':
                    game_state = GameState(game_state.bankroll, float(clause[1:]), game_state.round_num)
                elif clause[0] == 'P':
                    active = int(float(clause[1:]))
                elif clause[0] == 'H':
                    hands = [[], []]
                    hands[active] = clause[1:].split(',')
                    pips = [SMALL_BLIND, BIG_BLIND]
                    stacks = [STARTING_STACK - SMALL_BLIND, STARTING_STACK - BIG_BLIND]
                    round_state = RoundState(0, 0, pips, stacks, hands, None, [], None)
                elif clause[0] == 'G':
                    bounties = ['-1', '-1']
                    bounties[active] = clause[1:]
                    round_state = RoundState(round_state.button, round_state.street, round_state.pips,
                                             round_state.stacks,
                                             round_state.hands, bounties, round_state.deck, round_state.previous_state)
                    if round_flag:
                        self.pokerbot.handle_new_round(game_state, round_state, active)
                        round_flag = False
                elif clause[0] == 'F':
                    round_state = round_state.proceed(FoldAction())
                elif clause[0] == 'C':
                    round_state = round_state.proceed(CallAction())
                elif clause[0] == 'K':
                    round_state = round_state.proceed(CheckAction())
                elif clause[0] == 'R':
                    round_state = round_state.proceed(RaiseAction(int(float(clause[1:]))))
                elif clause[0] == 'B':
                    round_state = RoundState(round_state.button, round_state.street, round_state.pips,
                                             round_state.stacks,
                                             round_state.hands, round_state.bounties, clause[1:].split(','),
                                             round_state.previous_state)
                elif clause[0] == 'O':
                    round_state = round_state.previous_state
                    revised_hands = list(round_state.hands)
                    revised_hands[1 - active] = clause[1:].split(',')
                    round_state = RoundState(round_state.button, round_state.street, round_state.pips,
                                             round_state.stacks,
                                             revised_hands, round_state.bounties, round_state.deck,
                                             round_state.previous_state)
                    round_state = TerminalState([0, 0], None, round_state)
                elif clause[0] == 'D':
                    assert isinstance(round_state, TerminalState)
                    delta = int(float(clause[1:]))
                    deltas = [-delta, -delta]
                    deltas[active] = delta
                    round_state = TerminalState(deltas, None, round_state.previous_state)
                    game_state = GameState(game_state.bankroll + delta, game_state.game_clock, game_state.round_num)
                elif clause[0] == 'Y':
                    assert isinstance(round_state, TerminalState)
                    hero_hit_bounty, opponent_hit_bounty = (clause[1] == '1'), (clause[2] == '1')
                    if active == 1:
                        hero_hit_bounty, opponent_hit_bounty = opponent_hit_bounty, hero_hit_bounty
                    round_state = TerminalState(round_state.deltas, [hero_hit_bounty, opponent_hit_bounty],
                                                round_state.previous_state)
                    self.pokerbot.handle_round_over(game_state, round_state, active)
                    game_state = GameState(game_state.bankroll, game_state.game_clock, game_state.round_num + 1)
                    round_flag = True
                elif clause[0] == 'Q':
                    return
            if round_flag:
                self.send(CheckAction())
            else:
                assert active == round_state.button % 2
                action = self.pokerbot.get_action(game_state, round_state, active)
                self.send(action)


def parse_args():
    parser = argparse.ArgumentParser(prog='python3 bot_main.py')
    parser.add_argument('--host', type=str, default='localhost', help='Host to connect to')
    parser.add_argument('port', type=int, help='Port on host to connect to')
    return parser.parse_args()


def run_bot(pokerbot, args):
    assert isinstance(pokerbot, BaseBot)
    try:
        sock = socket.create_connection((args.host, args.port))
    except OSError:
        print('Could not connect to {}:{}'.format(args.host, args.port))
        return
    socketfile = sock.makefile('rw')
    client = EngineClient(pokerbot, socketfile)
    client.run()
    socketfile.close()
    sock.close()

File: public/rust/src/lib/engine/engine_client.rs
use std::io::{BufRead, BufReader, Write};
use std::net::TcpStream;
use clap::Parser;

use crate::lib::base::base_bot::BaseBot;
use crate::lib::game::poker_moves::PokerMove;
use crate::lib::game::poker_state::{GameState, RoundState, TerminalState};
use crate::lib::game::poker_state::{STARTING_STACK, BIG_BLIND, SMALL_BLIND};

pub struct EngineClient<T: BaseBot> {
    pokerbot: T,
    stream: TcpStream,
}

#[derive(Parser, Debug)]
#[command(author, version, about, long_about = None)]
pub struct Args {
    /// Host to connect to
    #[arg(long, default_value = "localhost")]
    pub host: String,

    /// Port to connect to
    pub port: u16,
}

impl<T: BaseBot> EngineClient<T> {
    pub fn new(pokerbot: T, stream: TcpStream) -> Self {
        EngineClient { pokerbot, stream }
    }

    pub fn send(&mut self, action: PokerMove) {
        let code = match action {
            PokerMove::Fold => "F".to_string(),
            PokerMove::Call => "C".to_string(),
            PokerMove::Check => "K".to_string(),
            PokerMove::Raise(amount) => format!("R{}", amount),
        };
        
        writeln!(self.stream, "{}", code).expect("Failed to write to socket");
    }

    pub fn run(&mut self) {
        let mut game_state = GameState {
            bankroll: 0,
            game_clock: 0.0,
            round_num: 1,
        };
        
        let mut round_state = None;
        let mut active = 0;
        let mut round_flag = true;
        
        let reader = BufReader::new(self.stream.try_clone().expect("Failed to clone stream"));
        
        for line in reader.lines() {
            let line = line.expect("Failed to read line");
            let packet: Vec<&str> = line.trim().split(' ').collect();
            
            if packet.is_empty() {
                continue;
            }
            
            for clause in packet {
                if clause.is_empty() {
                    continue;
                }
                
                let first_char = clause.chars().next().unwrap();
                let rest = &clause[1..];
                
                match first_char {
                    'T' => {
                        let time: f64 = rest.parse().expect("Failed to parse time");
                        game_state = GameState {
                            bankroll: game_state.bankroll,
                            game_clock: time,
                            round_num: game_state.round_num,
                        };
                    },
                    'P' => {
                        active = rest.parse::<f64>().expect("Failed to parse active") as usize;
                    },
                    'H' => {
                        let mut hands = [Vec::new(), Vec::new()];
                        hands[active] = rest.split(',').map(String::from).collect();
                        let pips = [SMALL_BLIND, BIG_BLIND];
                        let stacks = [STARTING_STACK - SMALL_BLIND, STARTING_STACK - BIG_BLIND];
                        
                        round_state = Some(RoundState {
                            button: 0,
                            street: 0,
                            pips,
                            stacks,
                            hands,
                            bounties: ["-1".to_string(), "-1".to_string()],
                            deck: Vec::new(),
                            previous_state: None,
                        });
                    },
                    'G' => {
                        if let Some(rs) = &mut round_state {
                            rs.bounties[active] = rest.to_string();
                            
                            if round_flag {
                                self.pokerbot.handle_new_round(&game_state, rs, active);
                                round_flag = false;
                            }
                        }
                    },
                    'F' => {
                        if let Some(rs) = &round_state {
                            if let Ok(new_rs) = rs.proceed(PokerMove::Fold) {
                                round_state = Some(new_rs);
                            }
                        }
                    },
                    'C' => {
                        if let Some(rs) = &round_state {
                            if let Ok(new_rs) = rs.proceed(PokerMove::Call) {
                                round_state = Some(new_rs);
                            }
                        }
                    },
                    'K' => {
                        if let Some(rs) = &round_state {
                            if let Ok(new_rs) = rs.proceed(PokerMove::Check) {
                                round_state = Some(new_rs);
                            }
                        }
                    },
                    'R' => {
                        if let Some(rs) = &round_state {
                            let amount: i32 = rest.parse().expect("Failed to parse raise amount");
                            if let Ok(new_rs) = rs.proceed(PokerMove::Raise(amount)) {
                                round_state = Some(new_rs);
                            }
                        }
                    },
                    'B' => {
                        if let Some(rs) = &mut round_state {
                            rs.deck = rest.split(',').map(String::from).collect();
                        }
                    },
                    'O' => {
                        if let Some(rs) = &mut round_state {
                            if let Some(prev_state) = &rs.previous_state {
                                let mut new_rs = (**prev_state).clone();
                                new_rs.hands[1 - active] = rest.split(',').map(String::from).collect();
                                
                                let terminal_state = TerminalState {
                                    deltas: [0, 0],
                                    bounty_hits: None,
                                    previous_state: Box::new(new_rs),
                                };
                                
                                // dont update round_state here because were transitioning to a terminal state
                            }
                        }
                    },
                    'D' => {
                        if let Some(rs) = &round_state {
                            let delta: i32 = rest.parse().expect("Failed to parse delta");
                            let mut deltas = [-delta, -delta];
                            deltas[active] = delta;
                            
                            game_state = GameState {
                                bankroll: game_state.bankroll + delta,
                                game_clock: game_state.game_clock,
                                round_num: game_state.round_num,
                            };
                        }
                    },
                    'Y' => {
                        if let Some(rs) = &round_state {
                            let chars: Vec<char> = rest.chars().collect();
                            let hero_hit_bounty = chars.get(0) == Some(&'1');
                            let opponent_hit_bounty = chars.get(1) == Some(&'1');
                            
                            let bounty_hits = if active == 1 {
                                [opponent_hit_bounty, hero_hit_bounty]
                            } else {
                                [hero_hit_bounty, opponent_hit_bounty]
                            };
                            
                            let terminal_state = TerminalState {
                                deltas: [0, 0], // should be set from D clause
                                bounty_hits: Some(bounty_hits),
                                previous_state: Box::new(rs.clone()),
                            };
                            
                            self.pokerbot.handle_round_over(&game_state, &terminal_state, active);
                            
                            game_state = GameState {
                                bankroll: game_state.bankroll,
                                game_clock: game_state.game_clock,
                                round_num: game_state.round_num + 1,
                            };
                            
                            round_flag = true;
                        }
                    },
                    'Q' => return,
                    _ => {}
                }
            }
            
            if round_flag {
                self.send(PokerMove::Check);
            } else if let Some(rs) = &round_state {
                assert_eq!(active, (rs.button % 2) as usize);
                let action = self.pokerbot.get_action(&game_state, rs, active);
                self.send(action);
            }
        }
    }
}

pub fn parse_args() -> Args {
    Args::parse()
}

pub fn run_bot<T: BaseBot>(pokerbot: T, args: Args) {
    match TcpStream::connect(format!("{}:{}", args.host, args.port)) {
        Ok(stream) => {
            let mut client = EngineClient::new(pokerbot, stream);
            client.run();
        },
        Err(_) => {
            eprintln!("Could not connect to {}:{}", args.host, args.port);
        }
    }
}